### ISSUE TITLE
[Feature] Sharing data between tests / Persistent Fixtures

### ISSUE THREAD

**Issue by arsinclair** [microsoft/playwright#24504]
Maybe I'm missing something and such feature already exists, if so, please point me in the right direction. I'm looking for a way to provide some data to my tests, the data will be fetched only once before running all tests and should then be passed to all tests. The data is fetched from the internet.

I tried using fixtures, but a fixture function runs every time a test runs, so that's an unnecessary overhead. I haven't found a way to persist my data in a JS object, the only workaround is to save it to a file, and then read the file in the fixture function.

I also tried using `test.beforeAll()` but somehow it still runs every single time a test is ran.

The test looks like that:

```ts
import { test, expect } from "@playwright/test";
import AllPages from "../../all-pages.json" assert { type: "json" };
import { calculateFileHash } from "../../utils/calculate-hash.js";
import { HashMap } from "../../utils/hash-map-manager.js";

test.describe("Test all pages", () => {
    for (const pageURL of AllPages) {
        test(`Visual Page Check for ${pageURL}`, async ({ page }) => {
            test.setTimeout(120000);

            await page.goto(pageURL, { waitUntil: "networkidle" });

            const screenshot = await page.screenshot({ fullPage: true });
            const screenshotHash = calculateFileHash(screenshot);
            const hashMapEntry = HashMap[screenshotHash];

            expect( /* expect the page screenshot hash to be in a hashmap */ );
        });
    }
});
```

The object I want to share between tests is the hashmap. Is this a functionality that's missing in Playwright, or I'm not looking in the right place? I've carefully scrutinised the docs beforehand.

**Comment by yury-s** [comment]
You can place the logic that fetches the state into a [global setup](https://playwright.dev/docs/test-global-setup-teardown#setup) project and have it run the data to a file. The tests may run in different [worker processes](https://playwright.dev/docs/test-parallel#worker-processes) and you cannot pass js object across process boundary, instead you need a [worker-scoped fixture](https://playwright.dev/docs/test-fixtures#worker-scoped-fixtures) that will read the file once per worker and have it available to the tests. Will that work for you?

**Comment by arsinclair** [comment]
So in ideal scenario the data should be read only once, and then persisted it in memory and shared with every test. Sort of a global configuration. I hear this is currently not something that Playwright can do, and I can imagine valid use cases for such approach. I'm curious if this is just an oversight or it's intentional. 

We want to store some metadata for the tests in the cloud and have Playwright pull it and use it when it runs. After it ran, this metadata is no longer needed. So in that sense, storing it in a file and then deleting the file after the suite has ran seems unnecessary.

If I correctly understand the workaround you suggested, the worker fixture will read the file once and then the results will be available to all tests executed by that worker. If so, I have two questions:
- for example, if I specify 4 workers in the configuration, I should expect that the file will be read 4 times for the entire test suite, or
- the worker's lifespan is shorter, and at some point Playwright might decide to shutdown a worker and spin up a new one

If we were to use this workaround, it is undesirable from a performance standpoint to read the file numerous times.

**Comment by yury-s** [comment]
> So in an ideal scenario the data should be read only once, and then persisted in memory and shared with every test. Sort of a global configuration. I hear this is currently not something that Playwright can do, but I can imagine valid use cases for such approach. I'm curious if this is just an oversight or it's intentional.

Your understanding is correct, we recommend using files for sharing data between tests/projects. The global setup project that I mentioned before will ensure that the metadata is fetched from the cloud only once before all tests and then in each worker runs tests requiring the data the file will be read only once in the new fixture.

> the worker's lifespan is shorter, and at some point Playwright might decide to shutdown a worker and spin up a new one

With `--workers 4` you tell Playwright to run at most 4 worker processes at the same time. It may shutdown a particular worker process and launch a new one due to various reasons, you can't count on there being exactly 4 processes and it should not matter in the grand scheme of things.

> If we were to use this workaround, it is undesirable from a performance standpoint to read the file numerous times.

Considering that when you run a playwright test Node.js already reads a lot of files and so does each browser when you launch it, reading one more file (which will also most likely be cached in RAM by the OS) in a fixture should have negligible impact on the overall performance. I'd not worry about it at all unless you see a measurable impact of that. So far we haven't heard any complaints about performance of such approach.
